docs/tutorial/writing-tests/test-case.rst

```diff
@@ -0,0 +1,1148 @@
+                64 bytes from 1.1.1.1: icmp_seq=1 ttl=53 time=39.0 ms
+                64 bytes from 1.1.1.1: icmp_seq=2 ttl=53 time=143 ms
+
+                --- 1.1.1.1 ping statistics ---
+                2 packets transmitted, 2 received, 0% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            network_available.network_available("wlan0", "90")
+            self.assertTrue(check_output_mock.called)
+
+        @mock.patch("subprocess.check_output")
+        def test_failure(self, check_output_mock):
+            check_output_mock.return_value = textwrap.dedent(
+                """
+                PING 1.1.1.1 (1.1.1.1) from 192.168.1.100 wlan0: 56(84) bytes
+                64 bytes from 1.1.1.1: icmp_seq=1 ttl=53 time=39.0 ms
+
+                --- 1.1.1.1 ping statistics ---
+                10 packets transmitted, a received, 90% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            with self.assertRaises(SystemExit):
+                network_available.network_available("wlan0", "0")
+
+.. note::
+   We use ``self.assertTrue(check_output_mock.called)`` instead of
+   ``check_output_mock.assert_called_once()``. The reason is that we have to be
+   compatible (in tests as well!) with python3.5 and
+   ``Mock.assert_called_once`` was introduced in Python3.6. If you don't know
+   when a function was introduced, refer to `the python documentation
+   <https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once>`_.
+   See below the example, if a function was introduced with a specific
+   version, you will find it there.
+
+To run the tests go to the root of the provider and run the following:
+
+.. code-block:: none
+
+    (checkbox_venv) > python3 manage.py test -u
+    test_failure (test_network_available.TestNetworkAvailable.test_failure) ...
+    [...]
+    test_nominal (test_network_available.TestNetworkAvailable.test_nominal) ...
+    [...]
+
+    ----------------------------------------------------------------------
+    Ran 2 tests in 0.002s
+
+    OK
+
+.. note::
+   You can also run ``python3 manage.py test`` without the ``-u``. Every
+   provider comes with a set of builtin tests like ``shellcheck``
+   (for the ``commands:`` sections) and flake8 (for all ``bin/*.py`` files).
+   Not providing ``-u`` will simply run all tests.
+
+Gathering Coverage from Unit Tests
+----------------------------------
+
+In Checkbox we have a coverage requirement for new pull requests.
+This is to ensure that new contributions do not add source paths that are not
+explored in testing and therefore easy to break down the line with any change.
+
+If you want to collect the coverage of your contribution you can run the
+following:
+
+.. code-block:: none
+
+    (checkbox_venv) > python3 -m coverage run manage.py test -u
```

Comment:
Looks like `coverage` has to be installed separately. so perhaps  `pip install coverage` is needed before this step